### PR TITLE
Solve #1786 implementation of onWebSocketClose

### PR DESCRIPTION
**Pull request by Inokinoki**
#1786 solved by adding a callback named onWebSocketClose. 
Thanks for the inspiring of @刘冰 at https://wiki.swoole.com/wiki/page/400.html. I find this elegant method for solving it (avoid the dependency between swoole_websocket_server and swoole_server).

Example for onWebSocketClose callback:
```php
$server = new swoole_websocket_server("0.0.0.0", 9502);

$server->on('close', function($server, $fd, $id) {
    echo "connection close: {$fd}\n";
});

$server->on('websocketclose', function($server, $code, $reason) {
    echo "connection close socket: {$code} Reason:{$reason}\n";
});
```

Test preparing

### CODE REVIEW

File: src/protocol/WebSocket.c [hunk 1]
```diff
@@ -318,6 +318,7 @@ int swWebSocket_dispatch_frame(swConnection *conn, char *data, uint32_t length)
             // Server attempt to close, frame sent by swoole_websocket_server->disconnect()
             conn->websocket_status = 0;
         }
+        swReactorThread_dispatch(conn, frame.str, length);
```
Comment (Inokinoki) [inline src/protocol/WebSocket.c]: Or we can move it to line 316. The dispatching to a closed connection takes no effect.

File: swoole_websocket_server.c [hunk 2]
```diff
@@ -415,6 +541,13 @@ static PHP_METHOD(swoole_websocket_server, on)
         sw_copy_to_stack(php_sw_server_callbacks[SW_SERVER_CB_onMessage], _php_sw_server_callbacks[SW_SERVER_CB_onMessage]);
         php_sw_server_caches[SW_SERVER_CB_onMessage] = func_cache;
     }
+    else if (strncasecmp("websocketclose", Z_STRVAL_P(event_name), Z_STRLEN_P(event_name)) == 0)
+    {
+        zend_update_property(swoole_websocket_server_class_entry_ptr, getThis(), ZEND_STRL("onWebSocketClose"), callback TSRMLS_CC);
+        php_sw_server_callbacks[SW_SERVER_CB_onWebSocketClose] = sw_zend_read_property(swoole_websocket_server_class_entry_ptr, getThis(), ZEND_STRL("onWebSocketClose"), 0 TSRMLS_CC);
+        sw_copy_to_stack(php_sw_server_callbacks[SW_SERVER_CB_onWebSocketClose], _php_sw_server_callbacks[SW_SERVER_CB_onWebSocketClose]);
+        php_sw_server_caches[SW_SERVER_CB_onWebSocketClose] = func_cache;
+    }
```
Comment (Inokinoki) [inline swoole_websocket_server.c]: Add websocketclose callback

File: swoole_websocket_server.c [hunk 3]
```diff
@@ -443,7 +576,7 @@ static PHP_METHOD(swoole_websocket_server, disconnect)
         RETURN_FALSE;
     }
 
-    if ((code < 4000 && code != 1000) || code >= 5000)
+    if ((code < 3000 && code != 1000) || code >= 5000)
```
Comment (Inokinoki) [inline swoole_websocket_server.c]: Expand the range of usable code in disconenct method

**Comment by twose** [comment]
补充一下单元测试

**Pull request merged**